server/src/main/java/io/crate/planner/optimizer/symbol/rule/SwapCastsInComparisonOperators.java

```diff
@@ -25,28 +25,46 @@
 import static io.crate.planner.optimizer.matcher.Pattern.typeOf;
 
 import java.util.List;
+import java.util.Objects;
 
 import io.crate.expression.scalar.cast.CastMode;
 import io.crate.expression.scalar.cast.ImplicitCastFunction;
 import io.crate.expression.symbol.Function;
+import io.crate.expression.symbol.Literal;
 import io.crate.expression.symbol.Symbol;
 import io.crate.expression.symbol.SymbolType;
 import io.crate.metadata.NodeContext;
+import io.crate.metadata.Reference;
 import io.crate.planner.optimizer.matcher.Capture;
 import io.crate.planner.optimizer.matcher.Captures;
 import io.crate.planner.optimizer.matcher.Pattern;
 import io.crate.planner.optimizer.symbol.FunctionLookup;
 import io.crate.planner.optimizer.symbol.Rule;
-import io.crate.types.ArrayType;
 import io.crate.types.DataType;
-import io.crate.types.DataTypes;
 
 
 public class SwapCastsInComparisonOperators implements Rule<Function> {
 
     private final Capture<Function> castCapture;
     private final Pattern<Function> pattern;
 
+    private static boolean isSwappable(Function cmpOp) {
+        Function cast = (Function) cmpOp.arguments().get(0);
+        if (cmpOp.arguments().get(1) instanceof Literal<?> literal) {
+            Reference ref = (Reference) cast.arguments().get(0);
+            DataType<?> refType = ref.valueType();
+            try {
+                var castedToRefType = refType.implicitCast(literal.value());
+                var castReverted = literal.valueType().implicitCast(castedToRefType);
+                return (Objects.equals(literal.value(), castReverted));
+            } catch (IllegalArgumentException | ClassCastException e) {
```

Comment:
Is there a possibility to use some method that reveals if the casts are possible, to avoid throwing and catching exceptions?

Reply:
Could not think of any good alternative. We must execute the casts to find out if `126::int`, or `128::int` can be casted to `byte` or not.

I tried to limit this logic to be applied only to number to number conversions that are `unsafe`.